### PR TITLE
Properties framework part 2 - gui component

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR introduces part 2 of the new properties framework, and exposes some more of the functionality of the framework to the GUI through QgsPropertyOverrideButton.

Basically this hooks up the transformer configuration and replaces the data defined assistants from QGIS 2.x. As well as porting the size/line width assistants from QGIS 2, it also brings new graphical assistants for mapping a value to a color from a color ramp and mapping values to a rotation output. 

These assistants are driven by the corresponding QgsPropertyDefinition for the linked property, which means that they're available consistently everywhere that properties are used. So the same line width assistant which is available for a simple line symbology layer is also available for eg diagram line thickness, and the color ramp based assistant can be used to set a label's background shape color...

As part of this the deprecated QgsDataDefined, QgsDataDefinedButton, QgsScaleExpression, and QgsSizeScaleWidget have all been removed. These have all been replaced by QgsProperty/QgsPropertyTransformers/QgsPropertyOverrideButton

### CODE REVIEW

**Comment by nyalldawson** [comment]
Screencast quickly running through some of this: https://www.youtube.com/watch?v=c35plhfI6M0&feature=youtu.be

**Comment by nirvn** [comment]
@nyalldawson , compiled just fine here, having tons of fun testing it out. The color ramp assistant for fill / stroke colors is brilliant.

**Comment by nirvn** [comment]
@nyalldawson , I've been testing this out yesterday afternoon as well as this morning. All seems to work as it should. IMO, the upgraded assistant's shortcut menu item should be moved to be at the top of the data defined / property button.

File: src/gui/qgspropertyassistantwidget.cpp [hunk 1]
```diff
@@ -416,15 +416,29 @@ QgsPropertyColorAssistantWidget::QgsPropertyColorAssistantWidget( QWidget* paren
...
   }
 
   connect( mNullColorButton, &QgsColorButton::colorChanged, this, &QgsPropertyColorAssistantWidget::widgetChanged );
   connect( mColorRampButton, &QgsColorRampButton::colorRampChanged, this, &QgsPropertyColorAssistantWidget::widgetChanged );
+
+  if ( !mColorRampButton->colorRamp() )
+  {
+    // set a default ramp
+    QString defaultRampName = QgsProject::instance()->readEntry( QStringLiteral( "DefaultStyles" ), QStringLiteral( "/ColorRamp" ), QString() );
+    std::unique_ptr< QgsColorRamp > defaultRamp( QgsStyle::defaultStyle()->colorRamp( !defaultRampName.isEmpty() ? defaultRampName : QStringLiteral( "Blues" ) ) );
```
Comment (nirvn) [inline src/gui/qgspropertyassistantwidget.cpp]: @nyalldawson , what happens here if there are no default color ramp set, and no saved color ramp named "Blues" ?

Comment (nyalldawson) [inline src/gui/qgspropertyassistantwidget.cpp]: The check on the next line will catch this - the combo will just be initially blank with no ramp set.

**Comment by nirvn** [comment]
👍

**Pull request merged**